### DISCUSSION TITLE
Should wgpu types implement Send / Sync?

### DISCUSSION THREAD

**Discussion by Pauan** [gfx-rs/wgpu#4127] [General]
Right now pretty much all wgpu types implement both `Send` and `Sync`.

However, it has not been decided yet whether WebGPU will support multi-threading or not.

And even if it does support multi-threading, it has not been decided *which* types are thread-safe and which are not:

https://github.com/gpuweb/gpuweb/issues/354

That discussion uses web terminology, here is the Rust equivalent:

* "serializable" means the same behavior as `Arc`: the object is `Clone` + `Send` + `Sync`, it doesn't copy the data, it just sends the pointer to the other thread. And multiple threads can access the same object at the same time, so it is fully thread-safe and synchronized.

* "transferable" means `Send` + `!Sync`, which means you can *send* the object to another thread, but you can't *access* it from multiple threads at the same time.

* If an object is not "serializable" or "transferable" then that means `!Send` + `!Sync` (single-threaded only).

Note that the web semantics *does* matter for wgpu, because you can compile wgpu to Wasm + WebGPU, and Wasm can run on multiple threads in the browser. So wgpu has to match the multi-threaded behavior of the web.

`Device` will likely be `Send` + `Sync`, and other resources (like `Buffer`) will likely be `Send` + `Sync` as well, but that hasn't been decided yet.

If the wgpu types implement `Send` + `Sync`, then it is a breaking change to make them `Send` + `!Sync` in the future.

However, if the wgpu types implement `Send` + `!Sync`, then they can be changed to `Send` + `Sync` later without a breaking change.

At the very least, the  `CommandEncoder`, `ComputePassEncoder`, and `RenderPassEncoder` types (and associated types like `RenderPass`) should be `!Send` + `!Sync` since they will likely never be thread safe.

**Comment by Pauan** [comment]
From that discussion, I think there's a very high chance that most things will be `Send` + `Sync`, so the only real change that needs to happen is to make the `*Encoder` types `!Send` + `!Sync`.

**Comment by daxpedda** [comment]
I believe https://github.com/gfx-rs/wgpu/pull/3691 made everything on Web `!Send + !Sync`.
So this shouldn't be a problem until things change.

**Reply by Pauan** [reply]
That's a big footgun, because wgpu compiles to multiple backends. So somebody can spend months creating a game engine, and everything seems fine in Vulkan, but then when they try to compile it to Wasm everything breaks.

You should not be conditionally changing `Send` / `Sync` based on the compile target, that's a ***really*** bad idea.

**Reply by daxpedda** [reply]
Ah, you were suggesting to change it for native backends as well?

I agree that the situation isn't ideal, but on Web **right now** no types can be `Send` or `Sync`, so to make it equal all targets would have to be `!Send + !Sync`.

**Reply by Pauan** [reply]
> Ah, you were suggesting to change it for native backends as well?

Yes. One of the massive benefits of wgpu is that it's portable, so its behavior should be the same on every backend and platform.

> but on Web right now no types can be `Send` or `Sync`, so to make it equal all targets would have to be `!Send` + `!Sync`.

I'm not sure if that's the only solution, WebGL can run on multiple workers, and WebGPU can run on multiple workers too.

So in theory it should be possible to run wgpu on multiple workers on the web. It would require some big changes in how the WebGL / WebGPU backends work.

I'm imagining something like this: if atomics are enabled, then the WebGL / WebGPU backends would behave differently.

There would be a `static` pool of resources which is shared by all threads/workers. This pool would contain buffers/textures/etc. except it only contains Rust types like `Vec<u8>`, it doesn't contain any JS objects.

When the user does something like loading a texture, or writing to a buffer, it would just update the `Vec<u8>` in that `static`.

And then each thread/worker would use `thread_local!` to create a worker-specific pool. This thread-local pool would contain the actual JS objects (such as `GPUBuffer`, `WebGLTexture`, etc.)

When it tries to render a texture/buffer, it will check to see if the buffer/texture has been initialized in the thread-local pool. If it hasn't, then it initializes a new buffer/texture (by calling the JS methods to create a JS buffer/texture)

And then it copies the `Vec<u8>` data from the `static` resource into the thread-local resource.

So basically it's managing a separate JS copy of the buffer/texture per thread, and automatically copying the bytes from the `static` buffer/texture into the thread-local buffer/texture.

**Reply by Pauan** [reply]
If we're talking about *immediate* solutions, I see a couple:

1. Put in a `static_assertions::assert_cfg!(not(all(target_arch = "wasm32", target_feature = "atomics")));`

   This means that the types are `Send` + `Sync` on Wasm, but if you try to compile your wgpu + Wasm program with multi-threading then it gives a compile error.
   
2. Put in a runtime assertion that the wgpu code is only being run on 1 thread. So if the user tries to actually create wgpu types on multiple threads, then they get a runtime panic.

I think either of those solutions is preferable to the current status quo.

**Reply by daxpedda** [reply]
We have done something similar like you described in `Winit` to make the types `Send`/`Sync`.

---

>  1. Put in a `static_assertions::assert_cfg!(not(all(target_arch = "wasm32", target_feature = "atomics")));`
>     This means that the types are `Send` + `Sync` on Wasm, but if you try to compile your wgpu + Wasm program with multi-threading then it gives a compile error.

But this would basically mean that you can't use `wgpu` when you are using `atomics`.

I think we are going to disagree on this. I do agree that there are some significant drawbacks to having different `Send`/`Sync` per platform, but I prefer that compared to either removing `Send + Sync` on all other platforms or completely disabling `atomics` on Wasm.

We could have also enabled `Send + Sync` by default on Wasm and disabled it when you are using atomics, but the argument here was if you do that then all libraries depending on `wgpu` will probably not be compatible with `atomics`.

---

I'm not a maintainer here btw, just my opinion. But I'm sure actual Wgpu maintainers will be happy to discuss a solution like you outlined above to allow `Send + Sync` on Web. I'm even happy to work/help on a PR if somebody else does the legwork first, but I'm also happy to chime in as long as somebody else champions it.

**Reply by Pauan** [reply]
> But this would basically mean that you can't use wgpu when you are using atomics.

Yes, but that's already true today, since it's completely impossible for wgpu to share the underlying JS objects like `WebGLTexture`.

So if somebody tries to use wgpu with Wasm multi-threading, they'll get super bizarre runtime bugs, because the JS object pointers will be wrong / dangling.

JS objects are represented as a `u32` pointer into a heap. That heap lives in JS. When you create a JS object (such as WebGL objects, or WebGPU objects) it will put that JS object into the heap, and will then send a pointer to Rust.

However, every worker gets their own separate JS heap. Which means that if you try to send a JS object from one worker to another worker, the pointer will be wrong.

That's why `JsValue` is `!Send + !Sync`, to prevent that from happening. And the WebGL / WebGPU backends use JS objects, which means that the WebGL / WebGPU backends can never be `Send + Sync`.

So that means it's completely impossible to use wgpu on multiple threads in Wasm right now, *unless* you use my idea of having the `static` + `thread_local` pools and copying over the data.

**Reply by daxpedda** [reply]
> > But this would basically mean that you can't use wgpu when you are using atomics.
> 
> Yes, but that's already true today, since it's completely impossible for wgpu to share the underlying JS objects like `WebGLTexture`.

I think you are misunderstanding what I'm saying here, right now it is possible to use Wgpu with atomics, you just can't use it **multi**-threaded, but you can use it in another thread just fine.

As far as I understood, your proposal was to disable Wgpu entirely if you have atomics enabled.

**Reply by Pauan** [reply]
> As far as I understood, your proposal was to disable Wgpu entirely if you have atomics enabled.

That's why I had the second proposal of doing runtime checks to verify that the wgpu types are being run inside of a single thread.

That would allow each thread/worker to have its own separate copy of wgpu, but would disallow sharing between multiple workers.

**Reply by Pauan** [reply]
Oh and a third proposal, though I don't particularly like this one:

3. Make the types `Send` + `Sync` if atomics is disabled, but make them `!Send` + `!Sync` if atomics is enabled.

   This has the advantage that the types are the same between all platforms, *if* you use single-threaded Wasm.

   And then if you enable multi-threaded Wasm, it then switches to `!Send + !Sync`, which makes it clear that you're not allowed to actually send the types between different threads.

I think that's worse than doing a runtime check, but I include it for completeness.

**Reply by daxpedda** [reply]
> That's why I had the second proposal of doing runtime checks to verify that the wgpu types are being run inside of a single thread.

Ah, I see, I misunderstood that, I thought you meant a runtime check instead of compile time check to disable it entirely.

Yes, I guess that could work actually. It still has the tradeoff that `Send`/`Sync` markers are incorrect though and will make it difficult for library consumers. Personally I would still prefer to not implement `Send + Sync` as long it's just incorrect, but that's not really up to me.

> 3\. Make the types `Send` + `Sync` if atomics is disabled, but make them `!Send` + `!Sync` if atomics is enabled.

This is the current status quo with the `fragile-send-sync-non-atomic-wasm` crate feature.

---

I have to say that as annoying as it was for some libraries to consume Wgpu after that PR, it made them force to think about the fact that these types do not implement `Send + Sync`. Libraries which didn't use `fragile-send-sync-non-atomic-wasm` now are more compatible with the atomics feature then they were before and libraries that did use the crate feature are now clearly not supporting the atomics feature.

So I'm overall quite happy with the outcome.
Again, not ideal, but a net-positive in my book.

**Reply by Pauan** [reply]
> As far as I understood, your proposal was to disable Wgpu entirely if you have atomics enabled.

Actually, after thinking about it more, I think having a compile-time assertion for `atomics` is correct.

If the user enables atomics for Wasm, then that means it will use `SharedArrayBuffer` to share the memory between all the Workers.

However, they still can't use JS objects between multiple workers, because each worker has its own JS heap.

So you might say: "but what if they use a separate copy of wgpu on each worker?"

In that case they don't need to use atomics at all. They can just instantiate the Wasm separately on each worker, so each worker will have a separate `ArrayBuffer` memory.

So for the use case of running wgpu on a separate worker, you don't need atomics in order to do that. So it should be fine to disable wgpu entirely if atomics is enabled.

**Reply by Pauan** [reply]
I guess a counter-argument is: "what if the user has shared state in an `Arc<Mutex<T>>` that is being shared between all the workers, and they want to use that shared state to render with wgpu, but they still only run wgpu in one worker"...

**Reply by Pauan** [reply]
There is another way to make WebGL / WebGPU thread-safe...

First, we would change `Adapter` and `Surface` to be `!Send + !Sync`.

The `Surface` is tied to a particular `<canvas>` element (or an `OffscreenCanvas`), and those cannot be safely sent between threads.

The `Adapter` is also `!Send + !Sync` because it would contain all of the JS objects for that `Adapter` (buffers, textures, etc.)

So when the user creates an `Adapter`, it is tied to the thread where it was created.

In addition, the `Adapter` would contain an `Arc<Mutex<VecDeque<AdapterCommand>>>` which is a list of commands that should be executed on the `Adapter`.

Those commands would be things like "create a JS buffer, write data into a JS buffer, create a JS texture, etc."

Those commands are always executed on the same thread where the `Adapter` is, so that way it can access the JS objects that are inside of the `Adapter`.

When the user creates a `Device`, the `Device` would also contain the same `Arc<Mutex<VecDeque<AdapterCommand>>>` as the `Adapter`, so that way the `Device` can push commands into the adapter's queue.

And because `Device` (and the other types) are `Send + Sync`, that means you can send it to another thread and call all of the usual methods.

When the user calls a method on `Device` / `Texture` / `Buffer` / `Queue` / etc. it will push a new command into the `Arc<Mutex<VecDeque<AdapterCommand>>>`.

So basically it's proxying all of the methods to the thread where the `Adapter` was created. Since it's running on 1 thread, it can access the JS objects just fine.

There is precedence for this:

https://research.mozilla.org/2014/07/22/webgl-in-web-workers-today-and-faster-than-expected/

That uses a similar strategy to proxy WebGL commands from a Worker to the main thread.

Except we can do it much faster, because we can use `SharedArrayBuffer` and `Arc<Mutex<VecDeque<AdapterCommand>>>`, which means we don't need to use `postMessage`.

I think this is the best way to implement multi-threading with wgpu, since it means that all the textures / buffers / etc. only need to be created once.

**Reply by daxpedda** [reply]
This is exactly how we implemented `Send`/`Sync` on Winit.
It's also how I would implement it in Wgpu, it's just a lot of work :sweat_smile:!

**Reply by Pauan** [reply]
Alright, so I think the conclusion of this discussion is: just create a proper thread-safe implementation for wgpu when Wasm atomics are enabled. That resolves the entire problem perfectly.

And it doesn't matter whether WebGPU supports multi-threading or not, because the proxying implementation can be used even if WebGPU is single-threaded. So we can *always* impl `Send + Sync` regardless of what WebGPU decides to do, so it is future-proof.

**Reply by cwfitzgerald** [reply]
Okay, there's a lot to catch up on here.

I want to clarify something from the get-go, barring a wire implementation, the wgpu types will _always_ be `!Send` `!Sync` on wasm. The types have handles to a thread local js heap, so cannot be sent around. Even with sharable WebGPU types, those are explicitly passed to post-message and a new value is put together on the other end. The actual rust bytes are still pointing to thread locals (just thread locals that happen to refer to the same thing).

I'm quite apprehensive to write a wire implementation inside of wgpu's web backend. It doesn't really feel like the responsibility at our level in the chain. We've always been very upfront about our requirements to our users, and this includes things like never actually spawning threads, or doing anything like that, especially on web. There are arguments to be made for it, it's a larger discussion if it should be done in-tree, out of tree, or at all.

I agree things suddenly not being `Send`/`Sync` on wasm sucks, but that's the reality of the types we're looking with, and happens with all sorts of other abstraction libraries as well.

**Reply by daxpedda** [reply]
Personally I wouldn't mind a wire implementation inside `wgpu`, nothing lower ofc. This wouldn't require that Wgpu spawns any threads and I would go even further and say that it shouldn't even cause any kind of overhead if not used multi-threaded. Additionally it could be behind a crate feature as well.

That said, though we have managed to reduce the code complexity by a lot in Winit's wire implementation, it's basically non-intrusive, my experience so far with Wgpu tells me that it's probably going to be significantly more complicated. This is really my main concern here: I don't think this is worth it for Wgpu and I'm afraid the maintenance burden would be too big. Maybe this can also be nicely done out-of-tree after #3354, in which case I would argue that there is absolutely no reason to do a *wire implementation* in `wgpu` proper.

As I said before, as a heavy user of Wasm atomics, I'm overall quite satisfied with the status-quo and am looking forward when we actually get `Send`/`Sync` in Wasm to solve this problem properly.

(just my 2¢, I'm not a maintainer here!)

**Reply by Pauan** [reply]
> @cwfitzgerald I want to clarify something from the get-go, barring a wire implementation, the wgpu types will always be !Send !Sync on wasm. The types have handles to a thread local js heap, so cannot be sent around. Even with sharable WebGPU types, those are explicitly passed to post-message and a new value is put together on the other end. The actual rust bytes are still pointing to thread locals (just thread locals that happen to refer to the same thing).

That's true today, but that won't be true in the future. The WebGPU multi-threading explainer talk about accessing WebGPU types across Wasm threads:

https://github.com/gpuweb/gpuweb/wiki/The-Multi-Explainer#multi-threading-synchronous--webassembly

So the proxying implementation would just be temporary until we have proper WebGPU multithreading in Wasm.

Of course the WebGL backend would have to always use proxying, but it's legacy so it's to be expected that it uses some hacks.

> I'm quite apprehensive to write a wire implementation inside of wgpu's web backend.

Why is that? Having feature parity between all of the backends and platforms is clearly a good thing, since portability is one of wgpu's major strengths.

**Reply by daxpedda** [reply]
> That's true today, but that won't be true in the future. The WebGPU multi-threading explainer talk about accessing WebGPU types across Wasm threads:
> 
> https://github.com/gpuweb/gpuweb/wiki/The-Multi-Explainer#multi-threading-synchronous--webassembly

FWIW, I believe we are still pretty far away from **actually** `Send`/`Sync` in Wasm. We would need the [TC39 structs proposal](https://github.com/tc39/proposal-structs), which would first need to trickle down the whole tech-stack before it reaches us here in Wgpu.

All the other ideas I have encountered so far, like `postMessage()` or re-creating handles with IDs, isn't actually `Send`/`Sync`, but could of course be used to implement that better then a wire implementation.

**Reply by Pauan** [reply]
> FWIW, I believe we are still pretty far away from actually Send/Sync in Wasm. We would need the [TC39 structs proposal](https://github.com/tc39/proposal-structs), which would first need to trickle down the whole tech-stack before it reaches us here in Wgpu.

I don't think TC39 structs are relevant, since they don't solve the problem of needing to use `postMessage`.

The WebGPU types are already serializable (at least in theory), so they already behave like TC39 structs. You can send WebGPU types to other workers, you just have to use `postMessage` to do so.

The explainer gives some potential solutions that don't require `postMessage`.

**Reply by Pauan** [reply]
Also, it's still possible for the user to asynchronously share WebGPU types by manually using `postMessage`.

So we could have an API design like this:

```rust
// The to_js method converts the wgpu type into a JsValue which can be shared between workers by using postMessage
buffer.to_js()
```

```rust
// The try_from method attempts to convert the JsValue back into a wgpu type
Buffer::try_from(&js)
```

So that way users can call `to_js()` on the wgpu type, then use `postMessage` to send the JS value to another worker, and then use `try_from` to recreate the wgpu type from the JS value.

This is extra work for the user, but it would enable sharing wgpu types (at least with the WebGPU backend).

**Reply by cwfitzgerald** [reply]
> The WebGPU multi-threading explainer talk about accessing WebGPU types across Wasm threads:

That is true, I just am not holding my breath on any way for that to actually occur. All of the existing solutions have some downsides which will need to be debated in the webgpu group.

> Why is that? Having feature parity between all of the backends and platforms is clearly a good thing, since portability is one of wgpu's major strengths.

To be clear, I'm not saying we 100% shouldn't do this or anything. Mainly it's a lot more code to maintain and we try to stay away from these heavy handed polyfills, so the user gets exactly what they expect to happen.

> This is extra work for the user, but it would enable sharing wgpu types (at least with the WebGPU backend).

This is the api I was expecting us to expose, at least in the medium term.

**Reply by daxpedda** [reply]
> > FWIW, I believe we are still pretty far away from actually Send/Sync in Wasm. We would need the [TC39 structs proposal](https://github.com/tc39/proposal-structs), which would first need to trickle down the whole tech-stack before it reaches us here in Wgpu.
> 
> I don't think TC39 structs are relevant, since they don't solve the problem of needing to use `postMessage`.

Please correct me if I'm wrong, but the idea here is that TC39 structs should pave the way for either:
1. Allowing a storage space to store shared structs, which then could be passed around to every Web worker through `postMessage()`. This system could e.g. be used in the future by `wasm-bindgen` to allow for `Send`/`Sync` types.
2. Allow something like the [Wasm thread spawn proposal](https://github.com/abrown/thread-spawn) to use it in a more "integrated" way by decoupling Wasm multi-threading from Web workers. Ergo we won't have a Wasm instance per Web worker anymore.

**Reply by Pauan** [reply]
> Allowing a storage space to store shared structs, which then could be passed around to every Web worker through postMessage(). This system could e.g. be used in the future by wasm-bindgen to allow for Send/Sync types.

I think it depends on whether we can implement dynamic collections with them or not. Right now they only have a fixed-size `SharedFixedArray`, which limits how many objects you can share between threads.

So the TC39 struct proposal might not be enough to implement `Send` / `Sync`.

> Allow something like the [Wasm thread spawn proposal](https://github.com/abrown/thread-spawn) to use it in a more "integrated" way by decoupling Wasm multi-threading from Web workers. Ergo we won't have a Wasm instance per Web worker anymore.

Wasm threads don't change anything with regard to `Send` / `Sync`, the behavior is the same regardless of whether it's 1 Wasm instance or multiple Wasm instances

The fundamental issue is that we need a way to synchronously send a JS object from one worker/thread to another worker thread (i.e. not using `postMessage`).
